src/core/ext/transport/chttp2/transport/frame_data.cc

```diff
@@ -132,15 +131,16 @@ grpc_error* grpc_deframe_unprocessed_incoming_frames(
             p->is_frame_compressed = true; /* GPR_TRUE */
             break;
           default:
+            char* msg;
             gpr_asprintf(&msg, "Bad GRPC frame type 0x%02x", p->frame_type);
             p->error = GRPC_ERROR_CREATE_FROM_COPIED_STRING(msg);
             p->error = grpc_error_set_int(p->error, GRPC_ERROR_INT_STREAM_ID,
                                           static_cast<intptr_t>(s->id));
             gpr_free(msg);
-            msg = grpc_dump_slice(*slice, GPR_DUMP_HEX | GPR_DUMP_ASCII);
-            p->error = grpc_error_set_str(p->error, GRPC_ERROR_STR_RAW_BYTES,
-                                          grpc_slice_from_copied_string(msg));
-            gpr_free(msg);
+            p->error = grpc_error_set_str(
+                p->error, GRPC_ERROR_STR_RAW_BYTES,
+                grpc_slice_from_moved_string(grpc_core::UniquePtr<char>(
+                    grpc_dump_slice(*slice, GPR_DUMP_HEX | GPR_DUMP_ASCII))));
```

Comment:
Is the purpose of this change performance or avoiding a risk of leaks? If it's just the latter, then ignore this comment. If it's performance, is there any chance of it benefitting non-error code paths? Additionally, it's still leaving something on the table since grpc_dump_slice causes us to lose the info about the length of the contents, which grpc_slice_from_moved_string gets again by calling strlen. Can we perhaps avoid losing that information in grpc_dump_slice by also keeping around the information from the dump_out in gpr_dump?

Reply:
That's a good point. I added a version of `gpr_dump` that returns the length so that we can use it in this PR.
Using that method, I introduced `gpr_dump_slice_to_slice` which would (a) lower the boilerplate and (b) use the known len.